pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -456,44 +620,66 @@ impl InMemoryLayer {
         })
     }
 
-    // Write path.
+    /// Write path.
+    ///
+    /// Errors are not retryable, the [`InMemoryLayer`] must be discarded, and not be read from.
+    /// The reason why it's not retryable is that the [`EphemeralFile`] writes are not retryable.
+    /// TODO: it can be made retryable if we aborted the process on EphemeralFile write errors.
     pub async fn put_batch(
         &self,
         serialized_batch: SerializedBatch,
         ctx: &RequestContext,
-    ) -> Result<()> {
+    ) -> anyhow::Result<()> {
         let mut inner = self.inner.write().await;
         self.assert_writable();
 
-        let base_off = {
-            inner
-                .file
-                .write_raw(
-                    &serialized_batch.raw,
-                    &RequestContextBuilder::extend(ctx)
-                        .page_content_kind(PageContentKind::InMemoryLayer)
-                        .build(),
-                )
-                .await?
-        };
-
-        for SerializedBatchOffset {
-            key,
-            lsn,
-            offset: relative_off,
-        } in serialized_batch.offsets
-        {
-            let off = base_off + relative_off;
+        let base_offset = inner.file.len();
+
+        // Add the base_offset to the batch's index values which are relative to the batch start.
+        let index_values: Vec<SerializedBatchOffset> = serialized_batch
+            .offsets
+            .into_iter()
+            .map(|SerializedBatchOffset { key, lsn, value }| {
+                let InMemoryLayerIndexValueUnpacked {
+                    will_init,
+                    len,
+                    pos,
+                } = value.unpack();
+                anyhow::Ok(SerializedBatchOffset {
+                    key,
+                    lsn,
+                    value: InMemoryLayerIndexValue::new(InMemoryLayerIndexValueNewArgs {
+                        base_offset,
+                        batch_offset: pos,
+                        len: len.into_usize(),
+                        will_init,
+                    })?,
+                })
+            })
+            .collect::<anyhow::Result<Vec<_>>>()?;
```

Comment:
We're mapping a vector of SerializedBatchOffset into a vector of SerializedBatchOffset: shouldn't we modify in-place to avoid the allocation?  I think we're basically just adjusting the offsets here.

Reply:
fcb39d052477fd5f1431d1758a3e753a93d4c3ce

Reply:
Please re-review that commit

Reply:
LGTM